### PR TITLE
[hail] strftime, strptime

### PR DESCRIPTION
**Pull request by iitalics**
in this PR: almost working implementation of strftime/strptime functions. they can be called from `hail.experimental.time`. they convert between `tstr`'s and `tint64` seconds-since-epoch.

they don't completely work because the DateTimeFormatter needs to be supplied a timezone. a description of this timezone (also, the locale!) should be part of the IR. we should make the default behavior of the python functions be to determine the current locale/timezone and bake that into the IR; this way it doesnt matter what the locale is in the JVM's environment.

```
strftime(format, time) -> tstr
  format : tstr
  time : tint64

Display the given time (seconds since epoch) using the given C-like time format string.
```

```
strptime(string, format)
  string : tstr
  format : tstr

Parse the given time string using the given C-like time format string. Returns time as seconds since epoch.
```

### CODE REVIEW

**Comment by danking** [comment]
Rolled scorecard, came up John C.

**Comment by danking** [comment]
I guess we should really assign someone responsibility to own the PR and someone else as assignee. @johnc1231 Can you negotiate with compiler team to get someone responsible for this?

**Comment by johnc1231** [comment]
I'll try and clean up this PR today, then assign it to someone else to review.

**Comment by johnc1231** [comment]
Ah, less done than I thought it was, now reading the description and I realize we just haven't dealt with timezones yet.

**Comment by johnc1231** [comment]
Still have a couple more letters to test, but right now everything seems to be more or less working, with timezone passing now supported.

**Comment by danking** [comment]
huzzah! Great work John!

**Comment by johnc1231** [comment]
Ok, I think this is ready for review now, assigned to Patrick. I ended up explicitly requiring timezones as an argument because of a Java 8 bug that prevented our `strptime` function from automatically inferring them: https://bugs.openjdk.java.net/browse/JDK-8033662

I also currently just use the en_US locale for everything, I think that can be revisited in the future.

**Comment by johnc1231** [comment]
Tracking future work: https://github.com/hail-is/hail/milestone/9

File: hail/python/hail/experimental/time.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+from hail.expr.expressions.expression_typecheck import *
+from hail.expr.functions import _func
+from hail.typecheck import *
+
+@typecheck(format=expr_str,
+           time=expr_int64,
+           zone_id=expr_str)
+def strftime(format, time, zone_id):
+    """
+    Convert Unix timestamp to a formatted datetime string.
```
Comment (patrick-schultz) [inline hail/python/hail/experimental/time.py]: I think one or two examples would be helpful.

File: hail/python/hail/experimental/time.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        An id representing the timezone. See notes above.
+    """
+    return _func("strftime", hl.tstr, format, time, zone_id)
+
+@typecheck(time=expr_str,
+           format=expr_str,
+           zone_id=expr_str)
+def strptime(time, format, zone_id):
+    """
+    Interpret a formatted datetime string as a Unix timestamp.
```
Comment (patrick-schultz) [inline hail/python/hail/experimental/time.py]: Examples here too.

File: hail/python/hail/experimental/time.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+import hail as hl
+from hail.expr.expressions.expression_typecheck import *
+from hail.expr.functions import _func
+from hail.typecheck import *
+
+@typecheck(format=expr_str,
+           time=expr_int64,
+           zone_id=expr_str)
+def strftime(format, time, zone_id):
```
Comment (patrick-schultz) [inline hail/python/hail/experimental/time.py]: I find it a little confusing that `strftime` and `strptime` take their arguments in different orders. Was that copying from somewhere else? The python `strftime` is a method on a time object, and `strptime` takes the date string first, so the most consistent might be to take the time argument first in both cases.

Comment (johnc1231) [reply hail/python/hail/experimental/time.py]: I also think it's confusing. I was trying to copy python:

https://docs.python.org/3.7/library/time.html#time.strptime
https://docs.python.org/3.7/library/time.html#time.strftime

strftime has the format string first, strptime has it second.

Comment (patrick-schultz) [reply hail/python/hail/experimental/time.py]: Oh I missed the static functions. I guess that's so `strptime` can have a default format, and `strftime` can default to the current time.

Disregard, matching python is good.

**Comment by johnc1231** [comment]
Closing and opening my own so I don't have to reauthorize a lot

**Pull request closed without merging**